casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala

```diff
@@ -30,127 +30,164 @@ import monix.execution.Scheduler
 import scala.concurrent.duration.FiniteDuration
 import scala.util.Try
 
+class CasperInit[F[_]](
+    val conf: CasperConf,
+    val delay: FiniteDuration,
+    val runtimeManager: RuntimeManager[F]
+)
+
 object CasperPacketHandler extends CasperPacketHandlerInstances {
   implicit private val logSource: LogSource = LogSource(this.getClass)
 
   def apply[F[_]](implicit ev: CasperPacketHandler[F]): CasperPacketHandler[F] = ev
 
   def of[F[_]: LastApprovedBlock: Metrics: BlockStore: ConnectionsCell: NodeDiscovery: TransportLayer: ErrorHandler: RPConfAsk: SafetyOracle: Sync: Concurrent: Time: Log: MultiParentCasperRef: BlockDagStorage](
-      conf: CasperConf,
-      delay: FiniteDuration,
-      runtimeManager: RuntimeManager[F],
+      init: CasperInit[F],
       toTask: F[_] => Task[_]
   )(implicit scheduler: Scheduler): F[CasperPacketHandler[F]] =
-    BlockStore[F].getApprovedBlock.flatMap {
+    BlockStore[F].getApprovedBlock map {
       case Some(approvedBlock) =>
-        // We have an approved block so the network is already up and so no genesis ceremony needed.
-        val genesis = approvedBlock.candidate.flatMap(_.block).get
-        for {
-          validatorId <- ValidatorIdentity.fromConfig[F](conf)
-          casper <- MultiParentCasper.hashSetCasper[F](
-                     runtimeManager,
-                     validatorId,
-                     genesis,
-                     conf.shardId
-                   )
-          _                   <- MultiParentCasperRef[F].set(casper)
-          _                   <- Log[F].info("Making a transition to ApprovedBlockRecievedHandler state.")
-          abh                 = new ApprovedBlockReceivedHandler[F](casper, approvedBlock)
-          validator           <- Ref.of[F, CasperPacketHandlerInternal[F]](abh)
-          casperPacketHandler = new CasperPacketHandlerImpl[F](validator)
-        } yield casperPacketHandler
+        val msg    = "Found ApprovedBlock in storage, reconnecting to existing network"
+        val action = connectToExistingNetwork[F](approvedBlock, init)
+        (msg, action)
+      case None if (init.conf.approveGenesis) =>
```

Comment:
Guards in pattern matching disable the compiler check for exhaustiveness so I would rather not go this way